File: rewrite-javascript/rewrite/test/javascript/whitespace-attachment.test.ts
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {JavaScriptParser} from "../../src/javascript";
import {JavaScriptPrinter} from "../../src/javascript/print";
import {MarkerPrinter, PrintOutputCapture} from "../../src/print";
import {J} from "../../src/java";
import {Tree} from "../../src/tree";

function prettifyKind(kind: string): string {
    const match = kind.match(/\.([A-Z]+)\$(.+)$/);
    if (match) {
        return `${match[1]}.${match[2]}`;
    }
    return kind;
}


class OutputNode {
    constructor(
        public readonly element: Tree,
        public readonly children: (OutputNode | string)[] = []
    ) {}

    addChild(child: OutputNode | string): void {
        this.children.push(child);
    }

    toString(): string {
        const childrenStr = this.children.map(child => {
            if (typeof child === 'string') {
                return "Text(" + child + ")";
            } else {
                return child.toString();
            }
        }).join(', ');

        return `${prettifyKind(this.element.kind)}{${childrenStr}}`;
    }
}

/**
 * Custom PrintOutputCapture that builds a tree structure instead of a flat string
 */
class TreeStructurePrintOutputCapture extends PrintOutputCapture {
    public rootNodes: OutputNode[] = [];
    private nodeStack: OutputNode[] = [];

    constructor(markerPrinter: MarkerPrinter = MarkerPrinter.DEFAULT) {
        super(markerPrinter);
    }

    startNode(element: Tree): void {
        const node = new OutputNode(element);

        if (this.nodeStack.length > 0) {
            // Add to parent's children
            this.nodeStack[this.nodeStack.length - 1].addChild(node);
        } else {
            // This is a root node
            this.rootNodes.push(node);
        }

        this.nodeStack.push(node);
    }

    /**
     * Called when visiting ends for an LST element (afterSyntax)
     */
    endNode(): void {
        this.nodeStack.pop();
    }

    /**
     * Override append to add string content to the current node
     */
    override append(text: string | undefined): PrintOutputCapture {
        if (text && text.length > 0) {
            if (this.nodeStack.length > 0) {
                this.nodeStack[this.nodeStack.length - 1].addChild(text);
            }
            super.append(text);
        }
        return this;
    }

    /**
     * Get the tree structure as a formatted string
     */
    getTreeStructure(): string {
        return this.rootNodes.map(node => node.toString()).join('\n');
    }
}

/**
 * Custom JavaScriptPrinter that uses TreeStructurePrintOutputCapture
 */
class TreeCapturingJavaScriptPrinter extends JavaScriptPrinter {
    protected override async beforeSyntax(j: J, p: PrintOutputCapture): Promise<void> {
        if (p instanceof TreeStructurePrintOutputCapture) {
            p.startNode(j);
        }
        await super['beforeSyntax'](j, p);
    }

    protected override async afterSyntax(j: J, p: PrintOutputCapture): Promise<void> {
        await super['afterSyntax'](j, p);
        if (p instanceof TreeStructurePrintOutputCapture) {
            p.endNode();
        }
    }
}

function findWhitespaceViolations(rootNodes: OutputNode[]): string[] {
    const violations: string[] = [];

    function checkNode(node: OutputNode, path: string = 'root'): void {
        if (node.children.length > 0) {
            const firstChild = node.children[0];

            // Check if first child is a node (not text)
            if (firstChild instanceof OutputNode) {
                // Skip EnumValueSet - its children intentionally have their own prefix for uniform formatting
                if (node.element.kind === J.Kind.EnumValueSet) {
                    // Don't check whitespace attachment for EnumValueSet children
                } else if (firstChild.children.length > 0) {
                    // Check if the grandchild exists and is text with non-empty whitespace
                    const grandchild = firstChild.children[0];
                    if (typeof grandchild === 'string' && grandchild.trim() === '' && grandchild.length > 0) {
                        const parentKind = prettifyKind(node.element.kind);
                        const childKind = prettifyKind(firstChild.element.kind);
                        violations.push(`${parentKind} has child ${childKind} starting with whitespace |${grandchild}|. The whitespace should rather be attached to ${parentKind} around ${firstChild.toString()}.`);
                    }
                }
            }
        }

        // Recursively check all child nodes
        node.children.forEach((child, index) => {
            if (child instanceof OutputNode) {
                checkNode(child, `${path}[${index}]`);
            }
        });
    }

    rootNodes.forEach((node, index) => {
        checkNode(node, `root[${index}]`);
    });

    return violations;
}

describe('whitespace should be attached to the outermost element', () => {
    test.each([
        "const c =  function(): number { return 116; };",
        "const x = new Date();",
        "async function m(): void { await Promise.resolve(); }",
        "class ResponseHandler extends EventEmitter<{ success: string; error: Error }> {}",
        "import React = require('react');",
        `export const enum Result { Good = "Good", Bad = "Bad" }`,
        "for (let i = 1; i < [4, 3, 6].length; i++) {",
        `import "./rpc"; declare module "./tree" {}`,
        "const userScores = new Map<string, number>()",
        "function* generateUsers() { yield { id: 1 } };",
        "type T = undefined extends undefined ? string : never;",
        "const FirstEntity = class FirstEntityClass {};",
        `
        for (
          let i = 0; i < numberOfConnections; i++
        ) {}
        `
    ])('%s', async (sourceCode) => {
        // given
        const parser = new JavaScriptParser();
        const cu = await (await parser.parse({text: sourceCode, sourcePath: 'test.ts'}).next()).value;
        const capture = new TreeStructurePrintOutputCapture(MarkerPrinter.SANITIZED);
        const printer = new TreeCapturingJavaScriptPrinter();

        // when
        await printer.visit(cu, capture);

        // then
        const violations = findWhitespaceViolations(capture.rootNodes);
        expect(violations).toEqual([]);
    });
});

File: rewrite-python/rewrite/tests/python/whitespace_attachment_test.py
# Copyright 2026 the original author or authors.
# <p>
# Licensed under the Moderne Source Available License (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# <p>
# https://docs.moderne.io/licensing/moderne-source-available-license
# <p>
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import ast
from pathlib import Path
from typing import List, Optional, Union


from rewrite.java import J
from rewrite.python.support_types import Py
from rewrite.python.printer import PythonPrinter, PythonJavaPrinter, PrintOutputCapture
from rewrite.python._parser_visitor import ParserVisitor
from rewrite import Tree


def _prettify_type(tree: Tree) -> str:
    cls = type(tree)
    module = cls.__module__
    if 'python' in module:
        return f"Py.{cls.__name__}"
    elif 'java' in module:
        return f"J.{cls.__name__}"
    return cls.__name__


class OutputNode:
    def __init__(self, element: Tree):
        self.element = element
        self.children: List[Union['OutputNode', str]] = []

    def add_child(self, child: Union['OutputNode', str]) -> None:
        self.children.append(child)

    def __str__(self) -> str:
        children_str = ', '.join(
            f"Text({child})" if isinstance(child, str) else str(child)
            for child in self.children
        )
        return f"{_prettify_type(self.element)}{{{children_str}}}"


class TreeStructurePrintOutputCapture(PrintOutputCapture):
    def __init__(self):
        super().__init__()
        self.root_nodes: List[OutputNode] = []
        self._node_stack: List[OutputNode] = []

    def start_node(self, element: Tree) -> None:
        node = OutputNode(element)
        if self._node_stack:
            self._node_stack[-1].add_child(node)
        else:
            self.root_nodes.append(node)
        self._node_stack.append(node)

    def end_node(self) -> None:
        self._node_stack.pop()

    def append(self, text: Optional[str]) -> 'TreeStructurePrintOutputCapture':
        if text and len(text) > 0:
            if self._node_stack:
                self._node_stack[-1].add_child(text)
            super().append(text)
        return self


class TreeCapturingPythonJavaPrinter(PythonJavaPrinter):
    def _before_syntax(self, tree: J, p: PrintOutputCapture) -> None:
        if isinstance(p, TreeStructurePrintOutputCapture):
            p.start_node(tree)
        super()._before_syntax(tree, p)

    def _after_syntax(self, tree: J, p: PrintOutputCapture) -> None:
        super()._after_syntax(tree, p)
        if isinstance(p, TreeStructurePrintOutputCapture):
            p.end_node()


class TreeCapturingPythonPrinter(PythonPrinter):
    def __init__(self):
        super().__init__()
        self._delegate = TreeCapturingPythonJavaPrinter(self)

    def _before_syntax(self, tree: Union[Py, J], p: PrintOutputCapture) -> None:
        if isinstance(p, TreeStructurePrintOutputCapture):
            p.start_node(tree)
        super()._before_syntax(tree, p)

    def _after_syntax(self, tree: Union[Py, J], p: PrintOutputCapture) -> None:
        super()._after_syntax(tree, p)
        if isinstance(p, TreeStructurePrintOutputCapture):
            p.end_node()


def _find_whitespace_violations(root_nodes: List[OutputNode]) -> List[str]:
    violations: List[str] = []

    def check_node(node: OutputNode) -> None:
        if node.children:
            first_child = node.children[0]
            if isinstance(first_child, OutputNode):
                if first_child.children:
                    grandchild = first_child.children[0]
                    if isinstance(grandchild, str) and grandchild.strip() == '' and len(grandchild) > 0:
                        parent_kind = _prettify_type(node.element)
                        child_kind = _prettify_type(first_child.element)
                        violations.append(
                            f"{parent_kind} has child {child_kind} starting with whitespace "
                            f"|{grandchild}|. The whitespace should rather be attached to {parent_kind}."
                        )
        for child in node.children:
            if isinstance(child, OutputNode):
                check_node(child)

    for node in root_nodes:
        check_node(node)

    return violations


def _parse_python(source: str) -> Tree:
    source_path = Path("test.py")
    visitor = ParserVisitor(source, None, None)
    tree = ast.parse(source)
    cu = visitor.visit_Module(tree)
    return cu.replace(source_path=source_path)


def test_simple_assignment():
    # given
    source = "x = 1"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_function_definition():
    # given
    source = "def foo(x, y):\n    return x + y"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_dict_literal():
    # given
    source = "d = {'x': 1, 'y': 2}"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_class_definition():
    # given
    source = "class Foo:\n    def bar(self):\n        pass"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_import_statement():
    # given
    source = "from os.path import join"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_list_comprehension():
    # given
    source = "result = [x * 2 for x in range(10) if x > 3]"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_lambda_expression():
    # given
    source = "f = lambda x, y: x + y"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_multiline_function():
    # given
    source = "def greet(name):\n    greeting = 'Hello, ' + name\n    return greeting"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_try_except():
    # given
    source = "try:\n    x = 1\nexcept ValueError:\n    x = 0"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_for_loop():
    # given
    source = "for i in range(10):\n    print(i)"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []


def test_decorator():
    # given
    source = "@staticmethod\ndef foo():\n    pass"
    cu = _parse_python(source)
    capture = TreeStructurePrintOutputCapture()
    printer = TreeCapturingPythonPrinter()

    # when
    printer.print(cu, capture)

    # then
    assert capture.out == source
    violations = _find_whitespace_violations(capture.root_nodes)
    assert violations == []
